homeassistant/components/snapcast/media_player.py

```diff
@@ -114,227 +187,214 @@ class SnapcastGroupDevice(MediaPlayerEntity):
-        self.hass.data[DOMAIN][self._entry_id].clients.append(self)
+class SnapcastGroupDevice(SnapcastBaseDevice):
+    """Representation of a Snapcast group device."""
 
-    async def async_will_remove_from_hass(self) -> None:
-        """Disconnect client object when removed."""
-        self._client.set_callback(None)
-        self.hass.data[DOMAIN][self._entry_id].clients.remove(self)
+    _device: Snapgroup
 
-    def set_availability(self, available: bool) -> None:
-        """Set availability of group."""
-        self._attr_available = available
-        self.schedule_update_ha_state()
+    @classmethod
+    def get_unique_id(cls, host, id) -> str:
+        """Get a unique ID for a group."""
+        return f"{GROUP_PREFIX}{host}_{id}"
 
     @property
-    def identifier(self):
-        """Return the snapcast identifier."""
-        return self._client.identifier
+    def _current_group(self) -> Snapgroup:
+        """Return the group."""
+        return self._device
 
     @property
-    def name(self):
+    def name(self) -> str:
         """Return the name of the device."""
-        return f"{self._client.friendly_name} {CLIENT_SUFFIX}"
+        return f"{self._device.friendly_name} {GROUP_SUFFIX}"
 
     @property
-    def source(self):
-        """Return the current input source."""
-        return self._client.group.stream
+    def state(self) -> MediaPlayerState | None:
+        """Return the state of the player."""
+        if self.is_volume_muted:
+            return MediaPlayerState.IDLE
+        return STREAM_STATUS.get(self._device.stream_status)
+
+    async def async_set_latency(self, latency) -> None:
+        """Handle the set_latency service."""
+        raise ServiceValidationError("Latency can only be set for a Snapcast client.")
+
+    async def async_join(self, master) -> None:
+        """Handle the join service."""
+        raise ServiceValidationError("Entity is not a client. Can only join clients.")
+
+    async def async_unjoin(self) -> None:
+        """Handle the unjoin service."""
+        raise ServiceValidationError("Entity is not a client. Can only unjoin clients.")
+
+
+class SnapcastClientDevice(SnapcastBaseDevice):
+    """Representation of a Snapcast client device."""
+
+    _device: Snapclient
+
+    @classmethod
+    def get_unique_id(cls, host, id) -> str:
+        """Get a unique ID for a client."""
+        return f"{CLIENT_PREFIX}{host}_{id}"
 
     @property
-    def volume_level(self):
-        """Return the volume level."""
-        return self._client.volume / 100
+    def _current_group(self) -> Snapgroup:
+        """Return the group the client is associated with."""
+        return self._device.group
 
     @property
-    def is_volume_muted(self):
-        """Volume muted."""
-        return self._client.muted
+    def available(self) -> bool:
+        """Check device availability."""
+        return super().available and self._device.connected
```

Comment:
Please don't rely on the connected state for availability. Even if the device is turned off/disconnected the state can be read from the server and it can be controlled (e.q. you could change the source/volume/etc). Of course these changes take effect only when the device is connected again. So when the server is available all devices should be available, too. Disconnected devices then have the standby state: can be controlled, but are off.

Reply:
:+1:  Not my preference but understood.